<div class="card my-2">
  <div class="card-header">
    <a
      class="d-block collapsed" data-toggle="collapse" aria-expanded="true"
      id="files-heading"
      href="#files-card"
      aria-controls="files-card">
      Package file details
    </a>
  </div>

  <div id="files-card" class="card-body collapse" aria-labelledby="files-heading">
    <p class="card-text">
      This report was delivered to you with some other files which are detailed below.
    </p>

    <p class="card-text">
      As part of our standard analysis package, we have included some extra files that
      will let you further explore and interrogate your dataset.  These files include:
    </p>

    <ol class="list-group list-group-flush">
      <li class="list-group-item"><code>_report.html</code>: This analysis report.</li>
      <li class="list-group-item"><code>_markers.xlsx</code>: a list of marker genes corresponding to the
      clustering shown in this report.  Marker genes are genes that distinguish cells
      in one cluster from those in other clusters, but note that multiple clusters may
      share some markers and some clusters will not has as many (or any) marker genes
      as others.  For each marker gene,
      <ol>
        <li><code>AUROC</code> represents how specific the expression of this gene is to
      this cluster.</li>
        <li><code>avg_diff</code> represents the average expression of this gene within
      cells of this cluster versus cells outside this cluster.</li>
      </ol></li>
      <li class="list-group-item"><code>_.Rds</code>: an <code>Rdata</code> file which can be used to interact
      with your data via a web application called "CellView", which is detailed
      below.</li>
    </ol>


    {% if not adata.uns.get("is_aggregation", False) %}
    <p></p>
    <p class="card-text">
      The software used to generate count matrices from your sequencing data, 10X
      Genomics CellRanger, generates some preliminary analysis information which we
      have included.  These files include:
    </p>

    <ol class="list-group list-group-flush">
      <li class="list-group-item"><code>cloupe.cloupe</code>: an input file for 10X Genomic <a
    href="https://support.10xgenomics.com/single-cell-gene-expression/software/visualization/latest/what-is-loupe-cell-browser">Loupe
    Cell Browser</a>, an interactive tool for looking at clustering and gene
    expression within your dataset. Note that the clustering generated by 10X will
    differ from that shown above.  This is due to differences in QC, normalization, and
    clustering methods.
    </ol>
    {% endif %}
  </div>
</div>

<div class="card my-2">
  <div class="card-header">
    <a
      class="d-block collapsed" data-toggle="collapse" aria-expanded="true"
      id="software-heading"
      href="#software-card"
      aria-controls="software-card">
      Software used
    </a>
  </div>

  <div id="software-card" class="card-body collapse" aria-labelledby="software-heading">
    <table class="table table-sm mb-0">
    <caption>
      1. More information about 10X Genomics CellRanger, which is used for read alignment and
      count matrix generation, can be found
      <a href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/what-is-cell-ranger">here</a>.
      <br>
      2. Scanpy is a Python-based analysis package from the Fabian Theis lab; read their
      manuscript describing scanpy
      <a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-017-1382-0">here</a>.
    </caption>
    <thead>
    <tr>
      <th style="text-align: left" scope="col" colspan="1" rowspan="1">SampleID</th>
      <th style="text-align: right" scope="col" colspan="1" rowspan="1">10X CellRanger<sup>1</sup></th>
      <th style="text-align: right" scope="col" colspan="1" rowspan="1">10X Reference Genome<sup>1</sup></th>
      <th style="text-align: right" scope="col" colspan="1" rowspan="1">Scanpy<sup>2</sup></th>
    </tr>
  </thead>

  <tbody>
    {% if adata.uns.get("is_aggregation", False) %}

      {% for sampleid in adata.uns["sampleids"] %}
        <tr>
          <th scope="row">{{ sampleid }}</th>
          <td>{{ adata.uns["cellranger_version"][sampleid] }}</td>
          <td>{{ adata.uns["genome"][sampleid] }}, version {{ adata.uns["cellranger_reference_version"][sampleid] }}</td>
          <td>{{ scanpy_version }}</td>
        </tr>
      {% endfor %}

    {% else %}
      <tr>
        <th scope="row">{{ adata.uns["sampleid"] }}</th>
        <td>{{ adata.uns["cellranger_version"] }}</td>
        <td>{{ adata.uns["genome"] }}, version {{ adata.uns["cellranger_reference_version"] }}</td>
        <td>{{ scanpy_version }}</td>
      </tr>
    {% endif %}
    </tbody>
    </table>

  </div>
</div>
