<div class="row">
  <div class="col m-2 p-1 border border-dark rounded">
    <h2 class="title">QC procedure</h2>
    <div class="row">
      <div class="col">

        <p>
          Data was filtered in three passes.  During FASTQ generation, reads with more
          than 1 mismatch in the 8bp i7 index are excluded.  During alignment (using
          <a href="https://github.com/alexdobin/STAR">STAR</a>), only reads with MAPQ
          scores greater than 255 that are aligned to annotated transcripts are retained.
          Reads containing bases with Q30 scores below 3 are further excluded.
        </p>

        <p>
          After alignment, cell barcodes are filtered (up to 1 mismatch) against a
          whitelist of 737,500 barcodes provided by 10X Genomics.  Then barcodes
          associated with cells are distinguished from those associated with ambient mRNA
          using an adaptively computed UMI threshold of {{ adata.uns["10x_umi_cutoff"] }}
          counts.  More on how this is computed can be found in the <a
            href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/algorithms/overview#cell_calling">10X
            CellRanger documentation</a>.  After this filtering, a
          {{ adata.uns["raw_cells"] }} (cells) by {{ adata.uns["raw_genes"] }} (genes) counts
          matrix is generated.
        </p>
      </div>
    </div>

    <div class="row">
      <div class="col">
        {% for sampleid in adata.uns["sampleids"] %}
          <img class="mx-auto d-block"
            src="data:image/png;base64,{{ adata.uns['report_images']['qc']['violin'][sampleid] }}"
            alt="Violin plots of QC metrics."
            width="100%" />
        {% endfor %}
        <p>
          The raw counts matrix was lastly filtered by hand using the following criteria,
          also depicted above.  Cells that satisfy the following metrics were excluded:
        </p>
        <ul>
          <li>cells with fewer than {{ adata.uns["qc_cell_filter"]["threshold_n_genes"] }} genes</li>
          <li>cells with fewer than {{ adata.uns["qc_cell_filter"]["threshold_n_counts"] }} UMI counts</li>
          {% if "threshold_hemoglobin_counts" in adata.uns["qc_cell_filter"] %}
          <li>cells with greater than {{ adata.uns["qc_cell_filter"]["threshold_hemoglobin_counts"] }} raw hemoglobin
          gene counts (e.g. HBB or Hbb-bs)</li>
          {% endif %}
          {% if "threshold_percent_mito" in adata.uns["qc_cell_filter"] %}
          <li>cells where mitochondrial mRNA comprises more than {{ adata.uns["qc_cell_filter"]["threshold_percent_mito"] }}%
          of the total UMI counts</li>
          {% endif %}
          {% if "threshold_sequencing_saturation" in adata.uns["qc_cell_filter"] %}
          {% endif %}
        </ul>
        <p>
        and genes that satisfy the following per-gene metrics were excluded:
        </p>
        <ul>
          <li>genes expressed in fewer than {{ adata.uns["qc_gene_filter"]["threshold_n_cells"] }} cells</li>
          <li>genes with fewer than {{ adata.uns["qc_gene_filter"]["threshold_n_counts"] }} UMI counts</li>
        </ul>
      </div>
    </div>

    <div class="row">
      <div class="col">
        {% for sampleid in adata.uns["sampleids"] %}
          <img class="mx-auto d-block"
            src="data:image/png;base64,{{ adata.uns['report_images']['qc']['scatter'][sampleid] }}"
            alt="Gene vs UMI scatter plots colored by QC metrics."
            width="100%" />
        {% endfor %}
      </div>
    </div>

    <div class="row">
      <div class="col">
        <p>
          After QC, the filtered cell-gene matrix contains {{  adata.uns["qc_metrics"]["cells"] }}
          cells and {{ adata.uns["qc_metrics"]["genes"] }} genes.
          <ol>
            <li>{{ adata.uns["qc_metrics"]["low_count_genes_removed"] }} genes were removed due to too few counts.</li>
            <li>{{ adata.uns["qc_metrics"]["cells_removed"] }} cells were removed due to several factors.
            <ul>
              <li>{{ adata.uns["qc_metrics"]["low_count_cells_removed"] }} were removed due to low UMI and gene counts.</li>
              <li>{{ adata.uns["qc_metrics"]["red_blood_cells_removed"] }} were removed due to high counts of hemoglobin markers.</li>
              <li>{{ adata.uns["qc_metrics"]["high_mtrna_cells_removed"] }} were removed due to high mtRNA content.</li>
              <li>{{ adata.uns["qc_metrics"]["low_sequencing_saturation_cells_removed"] }} were removed due to low per-cell sequencing saturation.</li>
            </ul></li>
          </ol>
        </p>
      </div>
    </div>
  </div>
</div>
