<div class="row">
  <div class="col border border-dark rounded m-2 p-1">
    <div class="row">
      <div class="col">
        <h2 class="title">Package file details</h2>
        <p>
          This report was delivered to you with some other files which are detailed below.
        </p>

        <p>
          As part of our standard analysis package, we have included some extra files that
          will let you further explore and interrogate your dataset.  These files include:
        </p>
        <ol>
          <li><code>_report.html</code>: This analysis report.</li>
          <li><code>_markers.csv</code>: a list of marker genes corresponding to the
          clustering shown in this report.  Marker genes are genes that distinguish cells
          in one cluster from those in other clusters, but note that multiple clusters may
          share some markers and some clusters will not has as many (or any) marker genes
          as others.  For each marker gene,
          <ol>
            <li><code>AUROC</code> represents how specific the expression of this gene is to
          this cluster.</li>
            <li><code>avg_diff</code> represents the average expression of this gene within
          cells of this cluster versus cells outside this cluster.</li>
          </ol></li>
          <li><code>_.Rds</code>: an <code>Rdata</code> file which can be used to interact
          with your data via a web application called "CellView", which is detailed
          below.</li>
        </ol>

        <p>
          The software used to generate count matrices from your sequencing data, 10X
          Genomics CellRanger, generates some preliminary analysis information which we
          have included.  These files include:
        </p>

        <ol>
          <li><code>web_summary.html</code>: a page of metrics and other data recapitulating
        the metrics shown at the beginning of this report.</li>
          <li><code>cloupe.cloupe</code>: an input file for 10X Genomic <a
        href="https://support.10xgenomics.com/single-cell-gene-expression/software/visualization/latest/what-is-loupe-cell-browser">Loupe
        Cell Browser</a>, an interactive tool for looking at clustering and gene
        expression within your dataset. Note that the clustering generated by 10X will
        differ from that shown above.  This is due to differences in QC, normalization, and
        clustering methods.
        </ol>
      </div>
    </div>
    <div class="row">
      <div class="col">
        <h2 class="title">Software used</h2>
        <ol>
          <li><code>cellranger, version 2.2.0</code> [<a href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/what-is-cell-ranger">ref</a>]</li>
          <li><code>cellranger {{ adata.uns["genome"] }} reference, version 2.1.0</code> [<a href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/what-is-cell-ranger">ref</a>]</li>
          <li><code>scanpy, version 1.3.4</code> [<a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-017-1382-0">ref</a>]</li>
        </ol>
      </div>
    </div>
  </div>
</div>
